ui/src/funding/pool.ts

```diff
@@ -0,0 +1,335 @@
+        tx.wait();
+      })
+      .finally(loadPoolData);
+  }
+
+  async function getErc20Allowance(): Promise<BigNumberish> {
+    const account = getAccount();
+    if (account) {
+      return erc20TokenContract.allowance(
+        account.address,
+        contract.POOL_ADDRESS
+      );
+    } else {
+      return Promise.resolve(0);
+    }
+  }
+
+  async function approveErc20(): Promise<void> {
+    const unlimited = BigNumber.from(1).shl(256).sub(1);
+    return erc20TokenContract
+      .approve(contract.POOL_ADDRESS, unlimited)
+      .then((tx: ContractTransaction) => {
+        transaction.add(transaction.erc20Allowance(tx));
+        tx.wait();
+      })
+      .finally(loadPoolData);
+  }
+
+  return {
+    data,
+    getAccount,
+    onboard,
+    updateSettings,
+    topUp,
+    withdraw,
+    withdrawAll,
+    collect,
+    approveErc20,
+  };
+}
+
+/**
+ * Stores
+ */
+export const amountStore = svelteStore.writable("");
+export const budgetStore = svelteStore.writable("");
+export const receiverStore = svelteStore.writable("");
+
+/**
+ *
+ * Validations
+ *
+ */
+
+const constraints = {
+  // The constraints for a valid monthly contribution.
+  monthlyContribution: {
+    presence: {
+      message: "The amount is required",
+      allowEmpty: false,
+    },
+    numericality: {
+      strict: false,
+      greaterThan: -1,
+    },
+  },
+
+  // The constraints for a valid pool top up amount input.
+  topUpAmount: {
+    presence: {
+      message: "The amount is required",
+      allowEmpty: false,
+    },
+    numericality: {
+      strict: false,
+      greaterThan: 0,
+    },
+  },
+};
+
+export const receiverValidationStore = (): validation.ValidationStore => {
+  return validation.createValidationStore({}, [
+    {
+      promise: isAddress,
+      validationMessage: "Please provide a valid Ethereum address",
+    },
+  ]);
+};
+
+function isAddress(value: string): Promise<boolean> {
+  return Promise.resolve(ethers.utils.isAddress(value));
+}
+
+export const monthlyContributionValidationStore = (): validation.ValidationStore => {
+  return validation.createValidationStore(constraints.monthlyContribution);
+};
+
+// Validate a balance operation, either a 'Top Up' or a 'Cash out'.
+// The provided `balance` represents the account balance upon which
+// the value being validated will be compared for sufficiency.
+export const balanceValidationStore = (
+  balance: BigNumberish
+): validation.ValidationStore => {
+  return validation.createValidationStore(constraints.topUpAmount, [
+    {
+      promise: amount => Promise.resolve(BigNumber.from(balance).gte(amount)),
+      validationMessage: "Insufficient balance",
+    },
+  ]);
+};
+
+// Check whether the user has onboarded their pool.
+export function isOnboarded(data: PoolData): boolean {
+  return (
+    data.erc20Allowance > 0 &&
+    (data.receivers.size > 0 || data.amountPerBlock !== "0" || data.balance > 0)
```

Comment:
I’m not sure the two last comparisons work as intended because the left-hand side is `BigNumberish`. For example if it is a `BigNumber` that represents zero than the last comparison returns true (I think).

To address this I’d suggest making `amountPerBlock` and `balance` of type `BigNumber` and use the appropriate methods. In fact, everything that is of type `BigNumberish` should probably be of type `BigNumber`. This would reduce the number of cases that would need to be distinguished and would make it easier to handle such values.